lib/github_changelog_generator/gitlab_fetcher.rb

```diff
@@ -0,0 +1,449 @@
+
+    private
+
+    def stringify_keys_deep(indata)
+      case indata
+      when Array
+        indata.map do |value|
+          stringify_keys_deep(value)
+        end
+      when Hash
+        indata.each_with_object({}) do |(key, value), output|
+          output[key.to_s] = stringify_keys_deep(value)
+        end
+      else
+        indata
+      end
+    end
+
+    # Exception raised to warn about moved repositories.
+    MovedPermanentlyError = Class.new(RuntimeError)
+
+    def extract_request_args(args)
+      if args.size == 1 && args.first.is_a?(Hash)
+        args.delete_at(0)
+      elsif args.size > 1 && args.last.is_a?(Hash)
+        args.delete_at(args.length - 1)
+      else
+        {}
+      end
+    end
+
+    # This is wrapper with rescue block
+    #
+    # @return [Object] returns exactly the same, what you put in the block, but wrap it with begin-rescue block
+    def check_response
+      Retriable.retriable(retry_options) do
+        yield
+      end
+    rescue MovedPermanentlyError => e
+      fail_with_message(e, "The repository has moved, update your configuration")
+    rescue Gitlab::Error::Forbidden => e
+      fail_with_message(e, "Exceeded retry limit")
+    rescue Gitlab::Error::Unauthorized => e
+      fail_with_message(e, "Error: wrong GitHub token")
+    end
+
+    # Presents the exception, and the aborts with the message.
+    def fail_with_message(error, message)
+      Helper.log.error("#{error.class}: #{error.message}")
+      sys_abort(message)
+    end
+
+    # Exponential backoff
+    def retry_options
+      {
+        on: [Gitlab::Error::Forbidden],
+        tries: MAX_FORBIDDEN_RETRIES,
+        base_interval: sleep_base_interval,
+        multiplier: 1.0,
+        rand_factor: 0.0,
+        on_retry: retry_callback
+      }
+    end
+
+    def sleep_base_interval
+      1.0
+    end
+
+    def retry_callback
+      proc do |exception, try, elapsed_time, next_interval|
+        Helper.log.warn("RETRY - #{exception.class}: '#{exception.message}'")
+        Helper.log.warn("#{try} tries in #{elapsed_time} seconds and #{next_interval} seconds until the next try")
+        Helper.log.warn GH_RATE_LIMIT_EXCEEDED_MSG
+        Helper.log.warn @client.rate_limit
+      end
+    end
+
+    def sys_abort(msg)
+      abort(msg)
+    end
+
+    # Print specified line on the same string
+    #
+    # @param [String] log_string
+    def print_in_same_line(log_string)
+      print log_string + "\r"
+    end
+
+    # Print long line with spaces on same line to clear prev message
+    def print_empty_line
+      print_in_same_line("                                                                       ")
+    end
+
+    # Returns GitHub token. First try to use variable, provided by --token option,
+    # otherwise try to fetch it from CHANGELOG_GITHUB_TOKEN env variable.
+    #
+    # @return [String]
+    def fetch_github_token
+      env_var = @options[:token].presence || ENV["CHANGELOG_GITHUB_TOKEN"]
```

Comment:
Q: Are there good reasons for this to be called GITHUB?

If not, create a GitLab-related environment variable.

The method name could be refactored, though.

Reply:
The main reason for me was to change as little as possible to get it work. I can tidy this up now

Reply:
As for the name, how about `fetch_auth_token`?